File: src/utils/urlMetadata/defaultParser.js
const cheerio = require('cheerio');
const logger = require('../logger');

async function fetchUrlMetadata(url) {
    try {
        const controller = new AbortController();
        const timeoutId = setTimeout(() => controller.abort(), 4000); // 4秒超時
        
        const response = await fetch(url, {
            signal: controller.signal,
            headers: {
                // 偽裝成瀏覽器，否則有些網站會阻擋
                'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36'
            }
        });
        clearTimeout(timeoutId);
        
        if (!response.ok) return null;
        
        const contentType = response.headers.get('content-type');
        if (!contentType || !contentType.includes('text/html')) return null;

        const text = await response.text();
        const $ = cheerio.load(text);
        
        let title = $('meta[property="og:title"]').attr('content') || $('title').text() || '';
        let description = $('meta[property="og:description"]').attr('content') || $('meta[name="description"]').attr('content') || '';
        
        if (!title && !description) return null;

        return { title: title.trim(), description: description.trim() };
    } catch (err) {
        logger.warn(`抓取網址失敗 (${url}): ${err.message}`);
        return null;
    }
}

module.exports = {
    match: () => true, // 預設解析器匹配所有無法被其他解析器處理的網址
    parse: async (url) => {
        const metadata = await fetchUrlMetadata(url);
        if (!metadata) return null;
        return `[網頁連結: ${metadata.title} | ${metadata.description}]`;
    }
};

File: src/utils/urlMetadata/defaultParser.py
import requests
from bs4 import BeautifulSoup
from ..logger import logger

def match(url):
    return True

def parse(url):
    try:
        headers = {
            'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36'
        }
        response = requests.get(url, headers=headers, timeout=4)
        if not response.ok:
            return None
            
        content_type = response.headers.get('content-type', '')
        if 'text/html' not in content_type:
            return None
            
        soup = BeautifulSoup(response.text, 'html.parser')
        
        title = ''
        og_title = soup.find('meta', property='og:title')
        if og_title:
            title = og_title.get('content', '')
        if not title:
            title_tag = soup.find('title')
            title = title_tag.text if title_tag else ''
            
        description = ''
        og_desc = soup.find('meta', property='og:description')
        if og_desc:
            description = og_desc.get('content', '')
        if not description:
            meta_desc = soup.find('meta', attrs={'name': 'description'})
            description = meta_desc.get('content', '') if meta_desc else ''
            
        if not title and not description:
            return None
            
        return {
            'text': f"[網頁連結: {title.strip()} | {description.strip()}]",
            'images': []
        }
    except Exception as e:
        logger.warning(f"抓取網址失敗 ({url}): {str(e)}")
        return None
